### PR TITLE
Check md5 values by calculating value from file

### LINKED ISSUE #18: error for invalid checksum

### PR DESCRIPTION
**Pull request by grosscol**
I reorganized checksum specs into subdirectory so they were split up by class.  This is why it looks like all the specs were re-written.

There are two gnarly parts of this PR that need to be acknowledged or resolved before removing the work in progress label:
1. Using a private method of the sip instance, ```file_in_zip```, in order to get the bytes on which to calculate the checksum
2. Using an integration test to exercise the code path that actually handles retrieving and crunching on the bytes from the zip file.

The use of the private method could be handled by using the public ```extract``` function.  It was my first approach, but got ugly in a hurry with trying to handle paths and disk i/o.  It could also be split off into a subsequent issue of "make file_in_zip a public fuction" that would involve doing some sort of test for the function.

Using the integration test is brittle because a change to the SIP class could bork the test without actually screwing up the functionality of the validator.  I was specifically thinking about an alteration to the SIP class' initialize arguments.  That might not affect other tests that mock the SIP, but could die when actually instantiating a new SIP that gets used on a fixture on disk.

Closes #18

### CODE REVIEW

**Comment by aelkiss** [comment]
Yeah, the issue in general of how to feed files into validators that actually need to check the contents in a testable way is a little bit tricky, and I'd like to discuss this today.

Some of the relevant issues currently:

- A single validator can look at multiple files - this precludes passing a single filehandle or the contents of a single file.
- The list of files the validator needs to look at is idiosyncratic to each validator (checksums vs. ocr, etc), so the logic that produces the list of files the validator needs to look at would need to live in the validator.

I'm wondering if it would make sense to have the notion of a per-file validator with a way to determine if it should run on any given file. That has a few advantages:

- the sip or siprunner can take care of reading the file in any way it cares about and can better optimize access to individual files.
- it would be easy to test the validator by passing a filehandle or file contents

I don't envision we'd use that approach for images, but we can see what makes sense in terms of an interface to FITS/JHOVE/whatever. We'll probably want to extract them from the zip and run FITS once on the extracted directory. But we can see what makes sense.

We can discuss this more in the meeting.

**Comment by grosscol** [comment]
The validators are currently getting a SIP instance, which is all the files.  I think that's probably the least complicated way to set it up.  Passing individual files or lists of files seems like it would too tightly entangle SIP and Validator.

The issue I ran into here was getting the stream of bytes of a file that the Validator needed to investigate.  That's I/O dependent and trying to handle that mess in this validator seems inappropriate.  So extracting to a directory and mucking with that was a lot more overhead than I wanted. 

Having the SIP handle getting the stream of bytes seems like a separation that fits here.  Ideally, it would hand over an IO object to a validator no matter what.  If it's an empty IO, the validator should be able to proceed as normal.  A filesystem or IO error should be handled by the SIP currently.  We can further abstract away the messy IO part from the SIP later.

File: lib/ht_sip_validator/validator/checksums/expected_value.rb [hunk 1]
```diff
@@ -6,26 +6,39 @@
...
+      create_error(
+        validation_type: :expected_checksum_value,
+        human_message: "Checksum mismatch for #{filename}",
+        extras: { filename: filename }
+      )
     end
 
+    private 
+
+    def calculate_checksums(filenames)
```
Comment (aelkiss) [inline lib/ht_sip_validator/validator/checksums/expected_value.rb]: Looks like the intention is to not bother checking the files in `filenames`, but that isn't the actual behavior

**Comment by grosscol** [comment]
Minimal refactor to use the `each_file` method of sip.  Left that in a private method in the validator, since testing it requires mocking `each_file` and/or `Digest::MD5.digest`

This refactor works, but still requires an integration test to exercise the private function.  I'll see about refactoring this on Monday or at least mark it as no longer in progress.

**Comment by aelkiss** [comment]
Yeah, this should work, but it would probably be better to refactor it as a subclass of `HathiTrust::Validator::FileValidator` which will more easily allow you to test if it's actually properly computing the checksums.

File: lib/ht_sip_validator/validator/checksums/expected_value.rb (hunk 1 above)
Comment (grosscol) [inline lib/ht_sip_validator/validator/checksums/expected_value.rb]: Good catch.  Vestiges of trying to be clever.  I'll remove that argument.  I realized that the exempt filenames are not going to take much to calculate a digest on, so might as well just do it.

**Comment by grosscol** [comment]
Sure. How is FileValidator going to be driven?

```ruby
errors = Array.new
file_validator = MyFileValidator(demo_sip)
demo_sip.each_file do |filename, instream|
  if file_validator.should_validate? filename
    errors << file_validator.perform_file_validation(filename, instream)
  end
end
```

**Comment by aelkiss** [comment]
More or less. See https://github.com/hathitrust/ht_sip_validator/blob/integration_fix/lib/ht_sip_validator/sip_validator_runner.rb#L57
https://github.com/hathitrust/ht_sip_validator/blob/integration_fix/lib/ht_sip_validator/validator/file_validator.rb

(Needs to change a little bit to deal properly with dependencies on a per-file basis, but don't worry about that right now)

**Pull request merged**